### PR TITLE
feat(docs): document build secrets

### PR DESCRIPTION
**Pull request by SasSwart**
This PR provides documentation for envbuilder build secrets.
Documentation for how to use build secrets from Terraform in conjunction with Coder will be provided separately.

Note that the documentation refers to `ghcr.io/coder/envbuilder:latest`. That will only work once we release this and push build secret support to ghcr. For now, you can test this by building your own version of the envbuilder container and using that instead.

### CODE REVIEW

File: docs/build-secrets.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+```bash
+docker run --rm -d -p 5000:5000 --name Envbuilder-registry registry:2
+```
+
+Then, build an image based on this Dockerfile:
+
+```Dockerfile
+FROM alpine:latest
+
+RUN --mount=type=secret,id=FOO,env cat $FOO > /foo_secret_hash.txt
```
Comment (johnstcn) [inline docs/build-secrets.md:23]: The naming of the output file here is somewhat confusing -- won't this just be the content of the secret and not the hash?

File: docs/build-secrets.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+RUN --mount=type=secret,id=BAR,dst=/tmp/bar.secret cat /tmp/bar.secret > /bar_secret_hash.txt
+```
+using this command:
+```bash
+docker run -it --rm \
+    -e ENVBUILDER_BUILD_SECRETS='FOO=envbuilder-test-secret-foo,BAR=envbuilder-test-secret-bar' \
+    -e ENVBUILDER_INIT_SCRIPT='/bin/sh' \
+    -e ENVBUILDER_CACHE_REPO=$(docker inspect Envbuilder-registry | jq -r '.[].NetworkSettings.IPAddress'):5000/test-container \
```
Comment (johnstcn) [inline docs/build-secrets.md:31]: Oh this is neat! I'd just been setting `--network=host` for the registry and using `localhost:5000`.

File: docs/build-secrets.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    -e ENVBUILDER_PUSH_IMAGE=1 \
+    -v $PWD:/workspaces/empty \
+    ghcr.io/coder/Envbuilder:latest
```
Comment (johnstcn) [inline docs/build-secrets.md:34]:
```suggestion
    ghcr.io/coder/envbuilder:latest
```

File: docs/build-secrets.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+```Dockerfile
+FROM alpine:latest
+
+RUN --mount=type=secret,id=FOO,env cat $FOO > /foo_secret_hash.txt
+RUN --mount=type=secret,id=BAR,dst=/tmp/bar.secret cat /tmp/bar.secret > /bar_secret_hash.txt
+```
+using this command:
+```bash
+docker run -it --rm \
+    -e ENVBUILDER_BUILD_SECRETS='FOO=envbuilder-test-secret-foo,BAR=envbuilder-test-secret-bar' \
```
Comment (johnstcn) [inline docs/build-secrets.md:29]: What does `envbuilder-test-secret-foo` mean in this context?

File: docs/build-secrets.md (hunk 2 above)
Comment (SasSwart) [reply docs/build-secrets.md:31]: Thanks! I saw the `--network=host` thing in some of your example templates. I got this trick from @mafredri 🌮

File: docs/build-secrets.md (hunk 1 above)
Comment (SasSwart) [reply docs/build-secrets.md]: Fixed.

File: docs/build-secrets.md (hunk 4 above)
Comment (SasSwart) [reply docs/build-secrets.md]: I've updated these secret names and added a description of their purpose in the documentation

File: docs/build-secrets.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+```bash
+docker run --rm -d -p 5000:5000 --name envbuilder-registry registry:2
+```
+
+Then, prepare the files to build our container.
+```bash
+mkdir test-build-secrets
+cd test-build-secrets
+printf 'FROM alpine:latest\n\nRUN --mount=type=secret,id=TEST_BUILD_SECRET_A,env=TEST_BUILD_SECRET_A echo -n $TEST_BUILD_SECRET_A | sha256sum > /foo_secret_hash.txt\nRUN --mount=type=secret,id=TEST_BUILD_SECRET_B,dst=/tmp/bar.secret cat /tmp/bar.secret | sha256sum > /bar_secret_hash.txt\n' > Dockerfile
+printf '{"build": { "dockerfile": "Dockerfile"}}\n' > devcontainer.json
```
Comment (johnstcn) [inline docs/build-secrets.md:23]: suggestion: change these to use `cat << EOF > file` for readability

File: docs/build-secrets.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+```
+
+Inspect the Dockerfile and devcontainer.json files in the new directory.
+```bash
+cat devcontainer.json
+cat Dockerfile
+```
```
Comment (johnstcn) [inline docs/build-secrets.md:30]: I don't think is strictly necessary if we use EOF above

File: docs/build-secrets.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+cat /proc/self/environ | tr '\0' '\n'
+printenv
+```
+* The secrets were still useful during the build. The following commands show that the secrets had side effects inside the build, without remaining in the image:
+```bash
+echo -n "secret-foo" | sha256sum
+cat /foo_secret_hash.txt
+echo -n "secret-bar" | sha256sum
+cat /bar_secret_hash.txt
+```
```
Comment (johnstcn) [inline docs/build-secrets.md:58]: suggestion: add a comment with the expected outputs

File: docs/build-secrets.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+### Who should be able to access build secrets, when and where?
+The secure way to use build secrets with Envbuilder is to deny users access to the platform that hosts Envbuilder. Only grant access to the Envbuilder container once it has concluded its build, using a trusted non-platform channel like ssh or the coder agent running inside the container. Once control has been handed to such a runtime container process, Envbuilder will have cleared all secrets that it set from the container.
+
+Anyone with sufficient access to attach directly to the container (eg. using `kubectl`), will be able to read build secrets if they attach to the container before it has concluded its build. Anyone with sufficient access to the platform that hosts the Envbuilder container will also be able to read these build secrets from where the platform stores them. This is true for other build systems, and containerised software in general.
```
Comment (johnstcn) [inline docs/build-secrets.md:108]: Maybe lead with this:

> Anyone with sufficient access to attach directly to the container (eg. using `kubectl`), will be able to read build secrets if they attach to the container before it has concluded its build.

**Review by johnstcn — APPROVED** [review]
Some suggestions for clarity 👍

File: docs/build-secrets.md (hunk 6 above)
Comment (SasSwart) [reply docs/build-secrets.md]: Jip. Removed

File: docs/build-secrets.md (hunk 7 above)
Comment (SasSwart) [reply docs/build-secrets.md:91]: Good idea. done.

File: docs/build-secrets.md (hunk 8 above)
Comment (SasSwart) [reply docs/build-secrets.md:150]: Done.

**Pull request merged**